components/support/migration/src/main/java/mozilla/components/support/migration/session/StreamingSessionStoreParser.kt

```diff
@@ -0,0 +1,227 @@
+     *
+     * {
+     *   "windows" [..]
+     * }
+     */
+    private fun parseWindows(reader: JsonReader): SessionManager.Snapshot {
+        var sessions = listOf<Session>()
+        var selection = -1
+
+        reader.beginArray()
+        if (reader.hasNext()) {
+            reader.beginObject()
+            parseWindow(reader).apply {
+                sessions = first
+                selection = second
+            }
+            reader.endObject()
+        }
+        while (reader.hasNext()) {
+            reader.skipValue()
+        }
+        reader.endArray()
+
+        return SessionManager.Snapshot(
+            sessions.map { SessionManager.Snapshot.Item(it) },
+            selection
+        )
+    }
+
+    /**
+     * Parses a window object.
+     *
+     * {
+     *   "tabs": [..]     // Delegating to [parseTabs]
+     *   "selected": ..   // Determines selected tab.
+     *   [..]             // Ignoring all other properties
+     * }
+     */
+    private fun parseWindow(reader: JsonReader): Pair<List<Session>, Int> {
+        var tabs: List<Session> = emptyList()
+        var selection = -1
+
+        while (reader.hasNext()) {
+            when (reader.nextName()) {
+                "tabs" -> {
+                    tabs = parseTabs(reader)
+                }
+                "selected" -> {
+                    selection = reader.nextInt() - 1
+                }
+                else -> {
+                    reader.skipValue()
+                }
+            }
+        }
+
+        return Pair(tabs, selection)
+    }
+
+    /**
+     * Parses a tab array.
+     *
+     * [
+     *   ... // Delegates to [parseTab]
+     * ]
+     */
+    private fun parseTabs(reader: JsonReader): List<Session> {
+        val tabs = mutableListOf<Session>()
+
+        reader.beginArray()
+
+        while (reader.hasNext()) {
+            reader.beginObject()
+            parseTab(reader)?.let { session ->
+                tabs.add(session)
+            }
+            reader.endObject()
+        }
+
+        reader.endArray()
+
+        return tabs
+    }
+
+    /**
+     * Parses a tab object.
+     *
+     * {
+     *   "entries": [..] // Delegates to [parseEntries]
+     *   "index": ..     // Used to determine current entry in tab history.
+     *   ..              // Ignoring all other properties
+     * }
+     */
+    private fun parseTab(reader: JsonReader): Session? {
+        var entries: List<Session> = emptyList()
+        var index = -1
+
+        while (reader.hasNext()) {
+            when (reader.nextName()) {
+                "entries" -> {
+                    entries = parseEntries(reader)
+                }
+                "index" -> {
+                    index = reader.nextInt()
+                }
+                else -> {
+                    reader.skipValue()
+                }
+            }
+        }
+
+        return if (index < 1 || entries.size < index) {
```

Comment:
Isn't it strange, a tab with no entries? Well, not much you can do here, other than maybe report it...

Reply:
Yeah, since I didn't write the format and since it is not really documented, I wanted to be over-protective here.